File: core/src/com/unibo/s3/Main.java
package com.unibo.s3;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector2;
import com.unibo.s3.main_system.AbstractMainApplication;
import com.unibo.s3.main_system.modules.BasicModule;
import com.unibo.s3.main_system.modules.CommandModule;

import java.util.ArrayList;
import java.util.List;

import static com.unibo.s3.main_system.rendering.ScaleUtils.*;


public class Main extends AbstractMainApplication {
    private List<BasicModule> modules = new ArrayList<>();
    private InputMultiplexer inputMultiplexer;

    @Override
    public void create() {
        super.create();
        inputMultiplexer = new InputMultiplexer();
        addModules();

        this.modules.forEach(m -> {
            m.init(this);
            m.attachInputProcessors(inputMultiplexer);
        });

        this.inputMultiplexer.addProcessor(this);
        Gdx.input.setInputProcessor(inputMultiplexer);

    }

    private void addModules() {
        CommandModule cm = new CommandModule();
        cm.enable(true);
        modules.add(cm);
        cm.setTextRenderer(textBatch);
    }

    private void renderAxis(ShapeRenderer shapeRenderer) {
        final Color oldColor = shapeRenderer.getColor();
        final Vector2 worldCenter = new Vector2(0f, 0f);

        final float metersRadius = 1f;
        final float pixelRadius = metersToPixels(metersRadius);

        final float axisMetersLength = 50;
        final float axisPixelLenght = metersToPixels(axisMetersLength);

        /*draw world center*/
        shapeRenderer.setColor(Color.RED);
        shapeRenderer.circle(worldCenter.x, worldCenter.y, pixelRadius);

        /*draw x axis*/
        shapeRenderer.setColor(Color.BLUE);
        shapeRenderer.line(-axisPixelLenght, 0, axisPixelLenght, 0);


        /*draw y axis*/
        shapeRenderer.setColor(Color.GREEN);
        shapeRenderer.line(0, -axisPixelLenght, 0, axisPixelLenght);
        shapeRenderer.setColor(oldColor);

        /*draw mouse WORLD vs SCREEN position*/
        final Vector2 mouseScreenPos = new Vector2(Gdx.input.getX(), Gdx.input.getY());

        /*scale by 1.0 / pixelPerMeter --> box2D coords*/
        final Vector2 mouseWorldPos = screenToWorld(mouseScreenPos).cpy().scl( getMetersPerPixel());

        //shapeRenderer.circle(metersToPixels(mouseWorldPos.x),
        //        metersToPixels(mouseWorldPos.y), metersToPixels(3));

        textBatch.begin();

            /*Flip y !!!*/
        font.setColor(Color.ORANGE);
        font.draw(textBatch, "screen: " + mouseScreenPos, Gdx.graphics.getWidth() / 2,
                30f);

        font.setColor(Color.YELLOW);
        font.draw(textBatch, "world: "+mouseWorldPos, Gdx.graphics.getWidth() / 2,
                15f);

        textBatch.end();
    }

    private void renderingTest(ShapeRenderer shapeRenderer) {
        final Color oldColor = shapeRenderer.getColor();
        shapeRenderer.setColor(Color.YELLOW);
        shapeRenderer.setAutoShapeType(true);
        shapeRenderer.set(ShapeRenderer.ShapeType.Line);
        shapeRenderer.circle(0, 0, getPixelsPerMeter() * 2);
        shapeRenderer.setColor(Color.GREEN);
        shapeRenderer.rect(-getPixelsPerMeter(), -getPixelsPerMeter(),
                getPixelsPerMeter() * 2, getPixelsPerMeter() * 2);
        shapeRenderer.setColor(oldColor);

    }

    @Override
    protected void doRender() {
        renderAxis(shapeRenderer);
        renderingTest(shapeRenderer);

        modules.stream()
                .filter(BasicModule::isEnabled)
                .forEach(m -> m.render(shapeRenderer));
        modules.stream()
                .filter(BasicModule::isEnabled)
                .forEach(BasicModule::renderGui);
    }

    @Override
    protected void doUpdate(float delta) {
        modules.stream()
                .filter(BasicModule::isEnabled)
                .forEach(m -> m.update(delta));
    }

    @Override
    public void resize(int newWidth, int newHeight) {
        super.resize(newWidth, newHeight);
        modules.forEach(m -> m.resize(newWidth, newHeight));
    }
}

File: core/src/com/unibo/s3/Main.scala
package com.unibo.s3

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.{Gdx, InputMultiplexer, Preferences}
import com.kotcrab.vis.ui.VisUI
import com.unibo.s3.main_system.AbstractMainApplication
import com.unibo.s3.main_system.communication.Messages.ToggleViewDebug
import com.unibo.s3.main_system.communication.{GeneralActors, SystemManager}
import com.unibo.s3.main_system.game._
import com.unibo.s3.main_system.modules._

class Main extends AbstractMainApplication {
  private var modules = List[BasicModule]()
  private var inputMultiplexer: InputMultiplexer = _

  private var bootstrapModule: BootstrapModule = _
  private[this] var cm: MenuModule = _
  private[this] val prefsName = "GntPreferences"
  private[this] var prefs: Preferences =  _

  override def create(): Unit = {
    super.create()
    prefs = Gdx.app.getPreferences(prefsName)
    inputMultiplexer = new InputMultiplexer
    VisUI.load()

    addModules()

    modules.foreach(m => {
      m.init(this)
      m.attachInputProcessors(inputMultiplexer)
    })

    inputMultiplexer.addProcessor(this)
    Gdx.input.setInputProcessor(inputMultiplexer)
  }

  private def addModules() = {

    val master = new MasterModule()
    master.enable(false)

    val lighting = new LightingSystemModule()
    lighting.enable(false)

    var settings: Option[GameSettings] = None
    val cm = new MenuModule({
      case Start(guardsNum, thievesNum, simulation, mapDimension, mazeTypeMap) =>
        var mapType: MapType = null
        mazeTypeMap match {
          case true => mapType = Maze
          case _ => mapType = Rooms
        }
        settings = Option(GameSettings(guardsNumber = guardsNum,
          thievesNumber = thievesNum, mapSize = mapDimension, mapType = mapType))
        bootstrapModule.enable(true)
      case Pause(pause) => paused = pause
      case ViewDebug(debug) =>
        val gameRef = SystemManager.getLocalActor(GeneralActors.GAME_ACTOR)
        gameRef ! ToggleViewDebug(debug)
    })
    cm.enable(true)

    bootstrapModule = new BootstrapModule({
      case BootstrapOk() =>

      case BootstrapFailed(err) =>
        println(err)

      case UserAck() if settings.isDefined =>
        lighting.setup()
        master.initGame(settings.get)
        master.enable(true)
        lighting.enable(true)
        removeModule(bootstrapModule)
    })
    bootstrapModule.enable(false)


    modules :+= cm
    modules :+= bootstrapModule
    modules :+= master
    modules :+= lighting
  }

  private def removeModule(m: BasicModule) = {

    Gdx.app.postRunnable(new Runnable {
      override def run(): Unit = {
        m.cleanup()
        m.enable(false)
        modules = modules.filter(mod => !mod.equals(m))
      }
    })
  }

  def getCamera: OrthographicCamera = cam

  def getPrefs: Preferences = prefs

  override protected def doRender(): Unit = {
    val enabledModules = modules.filter(m => m.isEnabled)
    enabledModules.foreach(m => m.render(shapeRenderer))
  }

  override def doCustomRender(): Unit = {
    val enabledModules = modules.filter(m => m.isEnabled)
    enabledModules.foreach(m =>m.customRender())
    enabledModules.foreach(m => m.renderGui())
  }

  override def doUpdate(delta: Float): Unit =
    modules.filter(m => m.isEnabled).foreach(m => m.update(delta))

  override def resize(newWidth: Int, newHeight: Int): Unit = {
    super.resize(newWidth, newHeight)
    modules.foreach(m => m.resize(newWidth, newHeight))
  }
}